### PR TITLE
Review

### PR DESCRIPTION
**Pull request by arpita-biswas**

### CODE REVIEW

File: RearrangingCars.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        
+        StringBuilder str = new StringBuilder();
+        
+        //If the initial or expected rearrangement array is null, then throw an exception.
+        if(initial == null || expected == null){
+            throw new NullRearrangementsException("Either the initial rearrangement or the expected rearrangement is null");
+        }
+        
+        //If the initial and expected rearrangement array do not have same length, then throw an exception.
+        else if(initial.length != expected.length){
```
Comment (nmeabed) [inline RearrangingCars.java:32]: I think there is no need for else if(), else. some if statements will do the job.

**Review by nmeabed — COMMENTED** [review]
good solution, however I think that the comments are too much as some lines are really self descriptive :) +1

File: RearrangingCars.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        }
+        
+        //If the initial or expected rearrangement array is not a valid permutation of {0,...,array.length-1}, then throw an exception.
+        else if(!isValidPermutation(initial) || !isValidPermutation(expected)){
+            throw new InvalidRearrangementException("The initial or the expected rearrangement is invalid!");
+        }
+        
+        else{
+            //If the initial and expected rearrangement is empty (no slots), then nothing to do, simply return "".
+            if(initial.length == 0){
```
Comment (koska18) [inline RearrangingCars.java:43]: I think lines from 43-56 are not necessary as your algorithm will quickly detect that arrays are empty or equal. In loop in line 49 you have to iterate over the whole array and it takes time in every single case, but I think that equal arrays are not so common case (I mean in most cases this loop is a waste of time). break in line 52 would help a bit also.

File: RearrangingCars.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+     * @return String: A string containing a series of steps required to arrange the cars
+     *  according to the "expected" rearrangement, starting from the "initial" rearrangement.
+     * @throws RearrangementException 
+     */
+    public static String rearrangingSteps(int [] initial, int [] expected) throws RearrangementExceptions{
+        
+        StringBuilder str = new StringBuilder();
+        
+        //If the initial or expected rearrangement array is null, then throw an exception.
+        if(initial == null || expected == null){
```
Comment (koska18) [inline RearrangingCars.java:27]: I appreciate good data validation, but I'm not sure if it's necessary as we were told that we don't have to validate input data. It takes almost a half of your code and it's easy to miss the main algorithm in all those exceptions :)

File: Tests.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        expectedStepsString.add(steps2.toString());
+        
+        String resultSteps = RearrangingCars.rearrangingSteps(initial, expected);
+        
+        assertTrue(expectedStepsString.toString().contains(resultSteps));
+    } 
+    
+    //Test when two pairs of slots need to be swapped: 6 steps required
+    @Test
+    public void test_twoswaps() throws RearrangementExceptions {
```
Comment (koska18) [inline Tests.java:137]: I don't think that writing all combinations of possible solutions is a good approach to tests :) It is really hard to maintain it and after changing initial array a bit you have to rewrite all expected moves.

File: RearrangingCars.java [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+                    isEqual = false;
+                }
+            }
+            if(isEqual){
+                return str.toString();
+            }
+            else{
+                //Obtain the required series of steps and print them
+                ArrayList<Step> steps = rearrangingStepsHelper(initial, expected);
+                str.append(steps.toString());
```
Comment (koska18) [inline RearrangingCars.java:60]: I don't think the StringBulider it's needed. You can return just steps.toString() and "" in cases above.

File: RearrangingCars.java (hunk 3 above)
Comment (nmeabed) [reply RearrangingCars.java:27]: @koska18  
I didn't notice and made all of these validations too , gotta edit that  👍

File: RearrangementExceptions.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+     * serialVersionUID = -1991874018749639875L
+     */
+    private static final long serialVersionUID = -1991874018749639875L;
+    public RearrangementExceptions() {}
+    public RearrangementExceptions(String message)
+    {
+       super(message);
+    }
+}
+
```
Comment (NatalijaGucevska) [inline RearrangementExceptions.java:52]: Great idea for the exceptions. We can definitely put this idea in our final solution.

File: RearrangementExceptions.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
+package CodeU_Assignment6;
+
+class NullRearrangementsException extends RearrangementExceptions{
+
+    /**
+     * serialVersionUID = 6995452535841683530L
+     */
+    private static final long serialVersionUID = 6995452535841683530L;
```
Comment (NatalijaGucevska) [inline RearrangementExceptions.java:8]: Why is this useful and why do you have it in commented version?

File: RearrangingCars.java (hunk 3 above)
Comment (NatalijaGucevska) [reply RearrangingCars.java:27]: I think that you shouldn't remove this once it is done. You can probably move it to a separate method such as `validateArguments`.

File: RearrangingCars.java (hunk 5 above)
Comment (NatalijaGucevska) [reply RearrangingCars.java:60]: +1. The StringBuilder is not really necessary here.

File: RearrangingCars.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+     * that is, the index of the empty slot.
+     * 
+     * Expects the "inverseMapping_array" to be of length >= 1.
+     * 
+     * @param inverseMapping_array: integer array denoting the inverse mapping
+     * @return Integer: integer denoting index of empty slot
+     */
+    private static int findEmptySlot(int [] inverseMapping_array){
+        return inverseMapping_array[0];
+    }
```
Comment (NatalijaGucevska) [inline RearrangingCars.java:187]: I'm not sure that one line call deserves a separate method. The positive side is that the name is self explanatory and you don't need to add comments while using it.  However you use this method only once (correct me if I'm wrong) and you comment the call to the method as well. 
Finally I think that you should avoid unnecessary abstraction.

File: RearrangingCars.java [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    /** indices_needToRearrange(current, expected) returns a HashSet of indices which 
+     * differs in "current" and "expected" rearrangement.
+     * 
+     * Expects the "array" to represent a valid rearrangement.
+     * 
+     * @param current: an array representing current rearrangement of the cars
+     * @param expected: an array representing expected rearrangement of the cars
+     * @return HashSet<Integer>: Set of all indices which differs "current" and "expected" rearrangement.
+     */
+    private static HashSet<Integer> indices_needToRearrange(int [] current, int [] expected){
```
Comment (NatalijaGucevska) [inline RearrangingCars.java:198]: This is a minor thing but I wouldn't mix `camelCase` naming with `underscore` naming.

File: RearrangingCars.java [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+     * @param array: integer array denoting a rearrangement of cars in spaces
+     * @return int[]: integer array denoting the inverse mapping
+     */
+    private static int [] getInverseMapping(int [] array){
+        int [] inverseMapping_array = new int[array.length];
+        for(int i=0; i<array.length; i++){
+            inverseMapping_array[array[i]] = i;
+        }
+        return inverseMapping_array;
+    }
```
Comment (NatalijaGucevska) [inline RearrangingCars.java:175]: Good idea of using a reversed array. I used HashMap which is more costly in memory than using directly an array.

File: RearrangingCars.java [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+    
+    /** isValidPermutation(array) returns whether or not the "array" contains
+     *  a valid permutation of [0,1,...,n] where n+1 is the length of the array.
+     * 
+     * @param array: integer array denoting a rearrangement of cars in spaces
+     * @return Boolean: true if the permutation is valid
+     */
+    private static boolean isValidPermutation(int [] array){
+        boolean found [] = new boolean[array.length];
+        for(int i=0; i<array.length; i++){
```
Comment (NatalijaGucevska) [inline RearrangingCars.java:149]: I would suggest to put `array.length` in a local variable since you make this call multiple times.

File: RearrangingCars.java [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+     *     i) If "move_from" is SAME as "empty_slot" (current and expected both have the same empty slot index),
+     *        then, select a random index (where the car parked differs from expected) and make it "move_from".
+     *        Then, go to step (3).
+     *  3. Make a move from move_from index to empty_slot index, and update the current rearrangement.
+     *  4. Repeat until the current rearrangement becomes same as that of the expected rearrangement.
+     * 
+     * @param current: an array representing current rearrangement of the cars
+     * @param expected: an array representing expected rearrangement of the cars
+     * @return: ArrayList<Steps>: representing the required series of steps
+     */
```
Comment (NatalijaGucevska) [inline RearrangingCars.java:84]: Thank you for this documentation, it helps to understand the solution. However I would suggest to put the Solution part in a comment in the method since it gives too much information to the user for the internal implementation.

File: Step.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+        this.move_from = move_from;
+        this.move_to = move_to;
+    }
+    
+    @Override
+    public String toString(){
+        StringBuilder str = new StringBuilder();
+        str.append("move from "+this.move_from+" to "+this.move_to);
+        return str.toString();
+    }
```
Comment (NatalijaGucevska) [inline Step.java:21]: You don't need the string builder in this case. You will use less memory if you just do `return "move from "+this.move_from+" to "+this.move_to`

File: Step.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    
+    @Override
+    public boolean equals(Object o){
+        Step obj = (Step) o;
```
Comment (NatalijaGucevska) [inline Step.java:25]: I think that it is a good practice to first verify that your object is `instanceOf` the class `Step`.

File: RearrangingCars.java [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+            inverseMapping_current[0] = move_from;                               
+
+            //Remove the empty_slot from the set of "yetToRearrange" indices, since it is occupied now
+            yetToRearrange.remove(empty_slot);
+            
+            //The index from which the car has moved, now contains 0 (no car) and becomes the new empty_slot 
+            empty_slot = move_from;                
+        }
+        return steps;
+    }
```
Comment (NatalijaGucevska) [inline RearrangingCars.java:139]: Adding comment to every line is really not necessary. For example `//If there is no more rearrangement required, then break` is a candidate to be removed.

**Review by NatalijaGucevska — COMMENTED** [review]
Great solution. I proposed minor changes. I don't think that you really need to maintain a list of cars that need to be moved  `yetToRearrange`. It is doing an unnecessary scan of the full array and you use additional memory to keep track of those cars. Such cars (that don't have to be moved) can be skipped by simple `if` condition such as `if(currentPosition != targetPosition)`.  I would be happy to hear your thoughts on this.

File: RearrangingCars.java [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+     *      3. Otherwise, call rearrangingStepsHelper() to get the required series of steps. 
+     *         Return the required steps as a String.           
+     * 
+     * @param initial: an array representing initial rearrangement of the cars
+     * @param expected: an array representing expected rearrangement of the cars
+     * @return String: A string containing a series of steps required to arrange the cars
+     *  according to the "expected" rearrangement, starting from the "initial" rearrangement.
+     * @throws RearrangementException 
+     */
+    public static String rearrangingSteps(int [] initial, int [] expected) throws RearrangementExceptions{
```
Comment (koska18) [inline RearrangingCars.java:22]: After working on Miruna's review I've noticed that she destroys `initial` array as she doesn't prepare a copy of it. You do the same thing, I suppose.

File: RearrangingCars.java (hunk 1 above)
Comment (arpita-biswas) [reply RearrangingCars.java:32]: Yes, you are right. :)

File: RearrangementExceptions.java (hunk 7 above)
Comment (arpita-biswas) [reply RearrangementExceptions.java:8]: This is not extremely crucial. However, I was reading a stackExchange post which says:

The serialization runtime associates with each serializable class a version number, called a serialVersionUID, which is used during deserialization to verify that the sender and receiver of a serialized object have loaded classes for that object that are compatible with respect to serialization. If the receiver has loaded a class for the object that has a different serialVersionUID than that of the corresponding sender's class, then deserialization will result in an  InvalidClassException. A serializable class can declare its own serialVersionUID explicitly by declaring a field named "serialVersionUID" that must be static, final, and of type long:

ANY-ACCESS-MODIFIER static final long serialVersionUID = 42L;
If a serializable class does not explicitly declare a serialVersionUID, then the serialization runtime will calculate a default serialVersionUID value for that class based on various aspects of the class, as described in the Java(TM) Object Serialization Specification. However, it is strongly recommended that all serializable classes explicitly declare serialVersionUID values, since the default serialVersionUID computation is highly sensitive to class details that may vary depending on compiler implementations, and can thus result in unexpected InvalidClassExceptions during deserialization. Therefore, to guarantee a consistent serialVersionUID value across different java compiler implementations, a serializable class must declare an explicit serialVersionUID value. It is also strongly advised that explicit serialVersionUID declarations use the private modifier where possible, since such declarations apply only to the immediately declaring class--serialVersionUID fields are not useful as inherited members.

Sorry for the long comment. I just copy pasted the answer from https://stackoverflow.com/questions/285793/what-is-a-serialversionuid-and-why-should-i-use-it

File: RearrangementExceptions.java (hunk 6 above)
Comment (arpita-biswas) [reply RearrangementExceptions.java:52]: Thanks :)

File: RearrangingCars.java (hunk 16 above)
Comment (arpita-biswas) [reply RearrangingCars.java:22]: Yes.

File: RearrangingCars.java (hunk 3 above)
Comment (arpita-biswas) [reply RearrangingCars.java:27]: I agree with @NatalijaGucevska . Let us keep these validations, however, we can put it in a separate method. 
I tried putting all the validations in the pre-processing and implemented the actual algorithm in the helper class.

File: RearrangingCars.java (hunk 15 above)
Comment (arpita-biswas) [reply RearrangingCars.java:139]: ok sure

File: Step.java (hunk 14 above)
Comment (arpita-biswas) [reply Step.java:25]: Ok

File: Step.java (hunk 13 above)
Comment (arpita-biswas) [reply Step.java:21]: Yes, sure.

File: RearrangingCars.java (hunk 11 above)
Comment (arpita-biswas) [reply RearrangingCars.java:149]: Thanks for the suggestion

File: RearrangingCars.java (hunk 3 above)
Comment (koska18) [reply RearrangingCars.java:27]: It's something we have to debate on :D

File: RearrangingCars.java (hunk 5 above)
Comment (arpita-biswas) [reply RearrangingCars.java:60]: I agree.

File: RearrangingCars.java (hunk 2 above)
Comment (arpita-biswas) [reply RearrangingCars.java:43]: I think it is necessary. The algorithm creates the inverse array in the first step and it takes O(n) memory and O(n) time. So it is better to pre-process it using O(n) time.

File: RearrangementExceptions.java (hunk 7 above)
Comment (NatalijaGucevska) [reply RearrangementExceptions.java:8]: Indeed, I have never payed attention that the Exception class was Serializable. Thanks for sharing this with me.

Comment (arpita-biswas) [reply RearrangementExceptions.java:8]: Neither I knew this, before doing this assignment :)

File: RearrangingCars.java (hunk 2 above)
Comment (koska18) [reply RearrangingCars.java:43]: I know but each of our approaches takes linear time and I think that there is a difference between 4n or 5n

Comment (arpita-biswas) [reply RearrangingCars.java:43]: Oh wait, I think I didn't get you. To reply to your comment "I think lines from 43-56 are *not* necessary as your algorithm will quickly detect that arrays are empty or equal"...
I said it is necessary to put those lines since a scan would take n time, whereas if I do not check (pre-process) it, then it might create an inverseArray which would take n space also along with n time. So it is better to pre-process it. We can discuss more in the evening meeting.

Comment (koska18) [reply RearrangingCars.java:43]: ok

File: RearrangingCars.java [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+     * @param current: an array representing current rearrangement of the cars
+     * @param expected: an array representing expected rearrangement of the cars
+     * @return: ArrayList<Steps>: representing the required series of steps
+     */
+    private static  ArrayList<Step> rearrangingStepsHelper(int [] current, int [] expected){
+        
+        ArrayList<Step> steps = new ArrayList<Step>();
+        
+        //Create an inverse-mapping of the "current" rearrangement.
+        int [] inverseMapping_current = getInverseMapping(current);
```
Comment (mirunamd) [inline RearrangingCars.java:90]: Just as a small comment, you can get around the need of having an inverse mapping array by using a HashMap as Natalija, Nada and I did because for me at least, it's a little bit hard to follow at the moment unless I have a pen and a paper in front of me
